rfcs/0049-flakes.md

```diff
@@ -0,0 +1,496 @@
+evaluation. We also need to disallow certain impurities that the Nix
+language previously allowed. In particular, the following are
+disallowed in a flake:
+
+* Access to files outside of the top-level flake or its inputs, as
+  well as paths fetched using `fetchTarball`, `fetchGit` and so on
+  without a commit hash or content hash. In particular this means that
+  Nixpkgs will not be able to use `~/.config/nixpkgs` anymore.
+
+* Access to the environment. This means that `builtins.getEnv "<var>"`
+  always returns an empty string.
+
+* Access to the system type (`builtins.currentSystem`).
+
+* Access to the current time (`builtins.currentTime`).
+
+* Use of the Nix search path (`<...>`); composition must be done
+  through flake inputs or `fetchX` builtins.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Pure evaluation breaks certain workflows. In particular, it breaks the
+use of the Nixpkgs configuration file. Similarly, there are people who
+rely on `$NIX_PATH` to pass configuration data to NixOps
+configurations.
+
+# Alternatives
+[alternatives]: #alternatives
+
+For composition of multi-repository projects, the main alternative is
+to continue on with explicit `fetchGit` / `fetchTarball` calls to pull
+in other repositories. However, since there is no explicit listing of
+dependencies, this does not provide automatic updating.
+
+Instead of a `flake.nix`, flakes could store their metadata in a
+simpler format such as JSON or TOML. This avoids the Turing tarpit
+where getting flake metadata requires the execution of an arbitrarily
+complex, possibly non-terminating program.
+
+Flakes could be implemented as an external tool on top of Nix. Indeed,
+there is nothing that flakes allow you to do that couldn't previously
+be done using `fetchGit`, the `--pure-eval` flag and some shell
+scripting. However, implementing flake-like functionality in an
+external tool would defeat the goals of this RFC. First, it probably
+wouldn't lead to a standard way to structure and compose Nix projects,
+since we might well end up with numerous competing
+"standards". Second, it would degrade rather than improve the Nix UX,
+since users would now have to deal with Nix *and* the flake-like tool
+on top of it.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+* Should flakes have arguments (like "system type")? This must be done
+  in a way that maintains hermetic evaluation and evaluation caching.
+
+* Currently, if flake dependencies (repositories or branches) get
+  deleted upstream, rebuilding the flake may fail. (This is similar to
+  `fetchurl` referencing a stale URL.) We need a command to gather all
+  flake dependencies and copy them somewhere else (possibly vendor
+  them into the repository of the calling flake).
+
+* Maybe flake metadata should be stored in a `flake.json` or
+  `flake.toml` file. This would prevent ambiguities when the Nix
+  language changes in a future edition.
```

Comment:
Sorry for being so late to the party, but while I understand that there were valid reasons to remove the `edition`-field in the end, wouldn't it make sense to keep a similar flag to make sure that `flake.*` files are interpreted properly even if we switch back and forth between e.g. `.nix`, `.toml` and sth else?

I'm asking because @edolstra stated in https://github.com/NixOS/rfcs/pull/49#issuecomment-637640007 that it's not 100% clear whether `.nix`-files are the way to go for flakes and I could imagine that we'd run into problems if we switch back to `.nix` with a different structure after several iterations.

Reply:
The decision was to name next version `flakes2.nix` or similar so there wouldn't be complex need of interpreting such files in a different manner.